x-pack/solutions/observability/plugins/streams/server/lib/streams/data_streams/manage_data_streams.ts

```diff
@@ -92,3 +93,55 @@ export async function rolloverDataStreamIfNecessary({
     }
   }
 }
+
+export async function updateDataStreamsLifecycle({
+  esClient,
+  logger,
+  names,
+  lifecycle,
+  isServerless,
+}: {
+  esClient: ElasticsearchClient;
+  logger: Logger;
+  names: string[];
+  lifecycle?: StreamLifecycle;
+  isServerless: boolean;
+}) {
+  try {
+    await retryTransientEsErrors(
+      () =>
+        esClient.indices.putDataLifecycle({
+          name: names,
+          data_retention: lifecycle?.type === 'dlm' ? lifecycle.data_retention : undefined,
+        }),
+      { logger }
+    );
+
+    // if we transition from ilm to dlm or vice versa, the rolled over backing
+    // indices need to be updated or they'll retain the lifecycle configuration
+    // set at the time of creation.
+    // this is not needed for serverless since only dlm is allowed.
+    if (isServerless) {
+      return;
+    }
+
+    const dataStreams = await esClient.indices.getDataStream({ name: names });
```

Comment:
note that this will still execute in stateful when updating a dlm (dlm -> dlm) while not necessary. I went the brute force way here but this can be optimized by filtering the backing indices not satisfying a condition. not sure if it's worth implemented now though

Reply:
Maybe leave this as a code comment as well?

Reply:
https://github.com/elastic/kibana/pull/207083/commits/e3f9dcc776af13e4e17d59f0c7dd2c7648a7e896